jgrapht-core/src/main/java/org/jgrapht/alg/GomoryHuCutTree.java

```diff
@@ -0,0 +1,157 @@
+                new DefaultDirectedWeightedGraph<V, DefaultWeightedEdge>(
+                                                                         DefaultWeightedEdge.class);
+
+            Graphs.addAllVertices(copy, graph.vertexSet());
+            for (E e : this.graph.edgeSet()) {
+                V v1 = this.graph.getEdgeSource(e);
+                V v2 = this.graph.getEdgeTarget(e);
+                Graphs.addEdge(copy, v1, v2, graph.getEdgeWeight(e));
+                Graphs.addEdge(copy, v2, v1, graph.getEdgeWeight(e));
+            }
+            return copy;
+        }
+
+    private void run(){
+        HashMap<V, V> predecessors = new HashMap<V, V>();
+        SimpleWeightedGraph<V, DefaultWeightedEdge> bogusGraph =
+            new SimpleWeightedGraph<V, DefaultWeightedEdge>(DefaultWeightedEdge.class);
+        DefaultDirectedWeightedGraph<V, DefaultWeightedEdge> directedGraph =
+            new DefaultDirectedWeightedGraph<V, DefaultWeightedEdge>(DefaultWeightedEdge.class);
+        directedGraph = this.makeDirectedCopy(this.graph);
+        MinSourceSinkCut minSourceSinkCut = new MinSourceSinkCut(directedGraph);
+
+        Graphs.addAllVertices(bogusGraph, this.graph.vertexSet());
+
+        Set<V> vertexSet = directedGraph.vertexSet();
+        Iterator<V> it = vertexSet.iterator();
+        V start = it.next();
+        predecessors.put(start, start);
+        while(it.hasNext()){
+            V vertex = it.next();
+            predecessors.put(vertex, start);
+        }
+
+        Iterator<V> itVertices = directedGraph.vertexSet().iterator();
+        itVertices.next();
+        while(itVertices.hasNext()){
+            V vertex = itVertices.next();
+            V predecessor = predecessors.get(vertex);
+            minSourceSinkCut.computeMinCut(vertex, predecessor);
+            Set<V> sourcePartition = minSourceSinkCut.getSourcePartition();
+            double flowValue = minSourceSinkCut.getCutWeight();
+            for(V sourceVertex : this.graph.vertexSet()){
+                if(!sourceVertex.equals(vertex)
+                   && predecessors.get(sourceVertex).equals(predecessor)
+                   && sourcePartition.contains(sourceVertex)){
+                    predecessors.put(sourceVertex, vertex);
+                }
+            }
+            if(sourcePartition.contains(predecessors.get(predecessor))){
+                V predecessorParent = predecessors.get(predecessor);
+                // p[s] = p[t]
+                predecessors.put(vertex, predecessorParent);
+                predecessors.put(predecessorParent, vertex);
+            }
+            DefaultWeightedEdge cutEdge =
+                Graphs.addEdge(bogusGraph, predecessor, vertex, flowValue);
+            cutEdges.add(cutEdge);
+        }
+    }
+
+    /**
+     * Returns the computed set of cut tree edges, not sorted after weight.
+     *
+     * @return The cut tree edges
+     */
+    public Set<DefaultWeightedEdge> getCutTreeEdges(){
```

Comment:
See notes above about what to return.  This will also allow the flawed test case to be fixed.